src/state.c

```diff
@@ -1265,6 +1265,27 @@ mod_mask_get_effective(struct xkb_keymap *keymap, xkb_mod_mask_t mods)
     return mask;
 }
 
+/* [WARNING] Caller must ensure that the modifier index is valid! */
+static inline bool
+is_real_mod(struct xkb_keymap *keymap, xkb_mod_index_t mod) {
+    return !!(keymap->mods.mods[mod].type & MOD_REAL);
```

Comment:
This check has to be done by the caller, as the comment states. My idea was to use a proper (inlined) function rather than a macro, for better typing/semantics. This is a very local function, but I could prefix it with `unsafe_` to highlight it.